### PR TITLE
0.2.4 Release - Various fixes

### LINKED ISSUE #8: Tree to String
Not sure if there's a way to easily accomplish this, but at present, if you specify a grouping for the source or destination, and attempt to convert it to a string, you are given (imo) unexpected results:

Test rule:
```
alert ip [127.0.0.1, 127.0.0.2] any -> ![8.8.8.8/24, 1.1.1.1] any (msg:"Test rule"; sid:12345678; rev:1;)
```

Example:
```
>>> import parsuricata
>>> test_rule = 'alert ip [10.10.10.0/24, !10.10.10.10] any -> ![8.8.8.8/24, 1.1.1.1] any (msg:"Test rule"; sid:12345678; rev:1;)'
>>> print(parsuricata.parse_rules(test_rule))
alert ip Tree(target_spec, [IPv4Interface('10.10.10.0/24'), Negated(value=IPv4Address('10.10.10.10'))]) any -> !Tree(target_spec, [IPv4Interface('8.8.8.8/24'), IPv4Address('1.1.1.1')]) any ( \
  msg: Test rule; \
  sid: 12345678; \
  rev: 1; \
)
```

I suppose I'd expect the string output to more accurately reflect the original rule. 

Appreciate any help you can provide, and also appreciate the work you put into this library (it is proving to be very helpful).

Thanks

### LINKED ISSUE #9: Error parsing individual port ranges
Reference: https://suricata.readthedocs.io/en/suricata-6.0.0/rules/intro.html#ports-source-and-destination

```parsuricata.parse_rules('''alert ip any 8000:9000 -> any 5000:6000 (msg:"Test rule"; sid:12345678; rev:1;)''')
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 62, in get_action
    return states[state][token.type]
KeyError: 'COLON'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jgrunzweig/work/tools/git/parsuricata/parsuricata/__init__.py", line 9, in parse_rules
    return parser.parse(source)
  File "/usr/local/lib/python3.8/site-packages/lark/lark.py", line 311, in parse
    return self.parser.parse(text, start=start)
  File "/usr/local/lib/python3.8/site-packages/lark/parser_frontends.py", line 89, in parse
    return self._parse(token_stream, start, *[sps] if sps is not NotImplemented else [])
  File "/usr/local/lib/python3.8/site-packages/lark/parser_frontends.py", line 54, in _parse
    return self.parser.parse(input, start, *args)
  File "/usr/local/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 36, in parse
    return self.parser.parse(*args)
  File "/usr/local/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 86, in parse
    action, arg = get_action(token)
  File "/usr/local/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 65, in get_action
    raise UnexpectedToken(token, expected, state=state)
lark.exceptions.UnexpectedToken: Unexpected token Token(COLON, ':') at line 1, column 18.
Expected one of: 
	* COMMA
	* RSQB
	* __ANON_6
	* LPAR
	* __ANON_7
```

Issue looks to be that, while port ranges are supported in groupings, they're not going to work by themselves. Should have a fix that'll tackle it. Going to submit a pull request shortly.

### LINKED ISSUE #10: Error parsing keywords with periods
Reference: https://suricata.readthedocs.io/en/suricata-6.0.0/rules/tls-keywords.html

```
>>> parsuricata.parse_rules('''alert ip any any -> any any (msg:"Test rule"; tls.cert_subject; content:"CN=*.googleusercontent.com"; sid:12345678; rev:1;)''')[0].__str__()
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/lark/lexer.py", line 373, in lex
    for x in l.lex(stream, self.root_lexer.newline_types, self.root_lexer.ignore_types):
  File "/usr/local/lib/python3.8/site-packages/lark/lexer.py", line 174, in lex
    raise UnexpectedCharacters(stream, line_ctr.char_pos, line_ctr.line, line_ctr.column, allowed=allowed, state=self.state, token_history=last_token and [last_token])
lark.exceptions.UnexpectedCharacters: No terminal defined for '.' at line 1 col 50

any any -> any any (msg:"Test rule"; tls.cert_subject; content:"CN=*.googleuserc
                                        ^

Expecting: {'SEMICOLON', 'COLON'}
```

Should be a simple fix of adding a period in the `KEYWORD` variable in grammar. Will include a fix for this in a pull request.

### PR DESCRIPTION
**Pull request by jgrunzweig**
Fixed for the following issues:
* https://github.com/theY4Kman/parsuricata/issues/8
* https://github.com/theY4Kman/parsuricata/issues/9
* https://github.com/theY4Kman/parsuricata/issues/10

### CODE REVIEW

**Comment by theY4Kman** [comment]
Oh, man, my GitHub notification game is shit, yo. I didn't even see this till tonight. Thank you for reporting these issues, and for finding fixes for them all! That's baller.

I would love to merge these ricky-tick, considering I left you hanging for 12-16 days. I do have a couple requests, though:

 - Would you take out the version number bump? I find it's best to keep version changes separate, so other PRs/changes can be packaged in, too; or so reverts (if needed) are easier.
 
 - Would you also change those commit messages to include descriptions of what they change? It's nice to be able to read the git history without having to reference GitHub. 
 
	Perhaps something like these:
	
	```
	fix: properly stringify parsed port groupings (fixes #8)
	fix: properly parse ungrouped port ranges (fixes #9)
	fix: allow periods in option keywords (fixes #10)
	```

I can do all those things directly on master if you'd rather not, but I'd really like to have yer face on the repo contributors ;) Looks like it's time for me to add a contributors section in the README (and jesus christ, a CHANGELOG).

Before I push any release up, I'm gonna add a few test cases to cover all these. In fact, I already wrote 'em to verify things :P
```py
EXAMPLE_RULES_CORPUS = '''
alert ip [127.0.0.1, 127.0.0.2] any -> ![8.8.8.8/24, 1.1.1.1] any ( msg:"Test rule"; sid:12345678; rev:1; )
alert ip any 80:100 -> any any ( msg:"Fart"; )
alert ip any 80: -> any any ( msg:"Fart"; )
alert ip any :100 -> any any ( msg:"Fart"; )
alert ip any any -> any any ( msg:"Test rule"; tls.cert_subject; content:"CN=*.googleusercontent.com"; sid:12345678; rev:1; )
'''

EXAMPLE_RULES = EXAMPLE_RULES_CORPUS.strip().splitlines()


@pytest.mark.parametrize('source_rule', [
    pytest.param(rule, id=rule)
    for rule in EXAMPLE_RULES
])
def test_parse_to_string_to_parse(source_rule):
    orig_parsed_rules = parse_rules(source_rule)
    orig_parsed_rule = orig_parsed_rules[0]

    stringified_rule = str(orig_parsed_rule)

    twice_parsed_rules = parse_rules(stringified_rule)
    twice_parsed_rule = twice_parsed_rules[0]

    expected = orig_parsed_rule
    actual = twice_parsed_rule
    assert expected == actual
```

(There's gotta be better names for `test_parse_to_string_to_parse`, `orig_parsed_rule`, and `twice_parsed_rule`, but brain no work right right now)

File: parsuricata/_parser.py [hunk 1]
```diff
@@ -71,12 +71,10 @@
 
     ?port_spec: variable
               | integer
-              | "[" port_grouping_spec ("," port_grouping_spec)* "]"    -> port_grouping
+              | port_range
+              | "[" port_spec ("," port_spec)* "]"                      -> port_grouping
               | "!" port_spec                                           -> negated
```
Comment (theY4Kman) [inline parsuricata/_parser.py:76]: These aliases really don't need to be indented so much, anymore. I think I've just been going by the rule "four spaces after the longest rule definition"
```suggestion
              | "[" port_spec ("," port_spec)* "]"    -> port_grouping
              | "!" port_spec                         -> negated
```

File: parsuricata/_parser.py [hunk 2]
```diff
@@ -91,7 +89,7 @@
     option: KEYWORD ";"
           | KEYWORD ":" settings ";"
 
-    KEYWORD: /[a-z_]+/i
+    KEYWORD: /[a-z_\.]+/i
```
Comment (theY4Kman) [inline parsuricata/_parser.py:92]: When inside a character group, the only thing that ever needs to be escaped is the closing bracket. That backslash can get goooone
```suggestion
    KEYWORD: /[a-z_.]+/i
```

File: parsuricata/__init__.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-__version__ = "0.2.3"
+__version__ = "0.2.4"
```
Comment (theY4Kman) [inline parsuricata/__init__.py:1]: tumblr_inline_mssg0qW1X91qz4rgp540
```suggestion
__version__ = "0.2.3"
```

**Review by theY4Kman — CHANGES_REQUESTED** [review]

**Comment by theY4Kman** [comment]
oh, hmm, I wonder if keywords are allowed to _start_ with a period...

**Comment by theY4Kman** [comment]
```
alert ip any [:100] -> any any ( msg:"end in group"; .stuff; sid: 8; )
```
```
3/6/2021 -- 00:53:11 - <Error> - [ERRCODE: SC_ERR_RULE_KEYWORD_UNKNOWN(102)] - unknown rule keyword '.stuff'.
```

Well, if _that's_ the error, I guess they _can_ technically start with a period.

File: parsuricata/__init__.py (hunk 3 above)
Comment (jgrunzweig) [reply parsuricata/__init__.py:1]: Sounds good to me

**Comment by theY4Kman** [comment]
I received an email today asking for support for option names with periods in them, so in the interest of expediency, I've gone and manually pushed these fixes to master. They've been published to PyPI as [version 0.2.4](https://pypi.org/project/parsuricata/0.2.4/).

**Pull request closed without merging**